### PR TITLE
Evm: Miner fixes

### PR DESCRIPTION
**Pull request by sieniven**
## Summary

- Fixes EVM block hash mismatch due to evm/transferdomain txs processed in different order on the miner and on connect block.
- Fixes adding incorrect txiter to failed tx set on miner pipeline that causes missing transferdomain txs inside evm block.
- Fixes segfault error on miner happening due to incorrect assertion from evm tx iters being processed again inside failedNonces set when they were already successfully minted into the block from another transferdomain tx iter descendant.

## Implications

- Storage
  - [ ] Database reindex required
  - [ ] Database reindex optional
  - [ ] Database reindex not required
  - [x] None

- Consensus
  - [ ] Network upgrade required
  - [ ] Includes backward compatible changes
  - [ ] Includes consensus workarounds
  - [ ] Includes consensus refactors
  - [x] None

### CODE REVIEW

File: test/functional/feature_evm_miner.py [hunk 1]
```diff
@@ -490,20 +489,158 @@ def same_nonce_transferdomain_and_evm_txs(self):
...
+                        "dst": {
+                            "address": self.ethAddress,
+                            "amount": "1@DFI",
+                            "domain": 3,
+                        },
+                        "nonce": start_nonce_erc55 + i,
+                    }
+                ]
+            )
+        self.nodes[0].generate(1)
```
Comment (kuegi) [inline test/functional/feature_evm_miner.py:606]: to recreate the crash: do not mine a block, send tx with the reused nonce while first batch is still in the mempool

Comment (sieniven) [reply test/functional/feature_evm_miner.py:606]: This issue should not be related to mempool but on the miner side.

Comment (kuegi) [reply test/functional/feature_evm_miner.py:606]: > This issue should not be related to mempool but on the miner side.

could be, but on my side I can not reproduce it if the block is already there. only happens when there is no UTXO left in the address and I reuse the nonce.

**Comment by kuegi** [comment]
FYI @sieniven on my side, if I change the part starting at line 518 to this, it fails (no block gets minted):

```

        for i in range(16):
            # tx call actual used gas - 1_761_626.
            # tx should always pass evm tx validation, but may return early in construct block.
            tx = contract.functions.loop(100_000).build_transaction(
                {
                    "chainId": self.nodes[0].w3.eth.chain_id,
                    "nonce": start_nonce + i,
                    "gasPrice": 25_000_000_000,
                    "gas": 30_000_000,
                }
            )
            signed = self.nodes[0].w3.eth.account.sign_transaction(tx, self.ethPrivKey)
            self.nodes[0].w3.eth.send_raw_transaction(signed.rawTransaction)

        amount= len(self.nodes[0].listunspent())+1
        for i in range(amount):
            self.nodes[0].transferdomain(
                [
                    {
                        "src": {
                            "address": self.address,
                            "amount": "1@DFI",
                            "domain": 2,
                        },
                        "dst": {
                            "address": self.ethAddress,
                            "amount": "1@DFI",
                            "domain": 3,
                        },
                        "nonce": start_nonce_erc55 + i,
                    }
                ]
            )

        start = self.nodes[0].getblockcount()
        self.nodes[0].generate(1)
        block_height = self.nodes[0].getblockcount()
        assert_equal(block_height, start + 1)
```

**Comment by kuegi** [comment]
Seems to be some randomness to my test. it works 50% of the runs, others fail.

File: src/miner.cpp [hunk 2]
```diff
@@ -873,12 +889,48 @@ void BlockAssembler::addPackageTxs(int& nPackagesSelected, int& nDescendantsUpda
...
             }
-            failedTxSet.insert(iter);
+
+            // Remove from checked TX set
+            for (const auto &entry : sortedEntries) {
+                checkedDfTxHashSet.erase(entry->GetTx().GetHash());
+            }
+
+            // Remove entries from queue if first EVM TX is not the failed TX.
+            if (sortedEntries.size() > 1) {
```
Comment (prasannavl) [inline src/miner.cpp:899]: Needed?

Comment (prasannavl) [reply src/miner.cpp:899]: Ok, never mind. Needed

File: src/miner.cpp [hunk 3]
```diff
@@ -873,15 +888,56 @@ void BlockAssembler::addPackageTxs(int& nPackagesSelected, int& nDescendantsUpda
...
+            for (const auto& entry : sortedEntries) {
+                auto entryTxType = entry->GetCustomTxType();
+                auto entryHash = entry->GetTx().GetHash();
+
+                if (entryTxType == CustomTxType::EvmTx || entryTxType == CustomTxType::TransferDomain) {
+                    // If the first TX in a failed set is not the failed TX
+                    // then remove from queue, otherwise it has not been added.
+                    if (entryHash != failedCustomTx) {
+                        CrossBoundaryResult result;
+                        auto hashes = evm_unsafe_try_remove_txs_above_hash_in_q(result, evmQueueId, entry->GetTx().GetHash().ToString());
```
Comment (prasannavl) [inline src/miner.cpp:911]:
```suggestion
                        auto hashes = evm_unsafe_try_remove_txs_above_hash_in_q(result, evmQueueId, entryHash.ToString());
```

File: src/miner.cpp [hunk 4]
```diff
@@ -873,15 +888,56 @@ void BlockAssembler::addPackageTxs(int& nPackagesSelected, int& nDescendantsUpda
...
+                auto entryHash = entry->GetTx().GetHash();
+
+                if (entryTxType == CustomTxType::EvmTx || entryTxType == CustomTxType::TransferDomain) {
+                    // If the first TX in a failed set is not the failed TX
+                    // then remove from queue, otherwise it has not been added.
+                    if (entryHash != failedCustomTx) {
+                        CrossBoundaryResult result;
+                        auto hashes = evm_unsafe_try_remove_txs_above_hash_in_q(result, evmQueueId, entryHash.ToString());
+                        if (!result.ok) {
+                            LogPrintf("%s: Unable to remove %s from queue. Will result in a block hash mismatch.\n", __func__, entry->GetTx().GetHash().ToString());
```
Comment (prasannavl) [inline src/miner.cpp:913]:
```suggestion
                            LogPrintf("%s: Unable to remove %s from queue. Will result in a block hash mismatch.\n", __func__, entryHash.ToString());
```

**Pull request merged**